File: backend/src/app/schemas/matching.py
"""Matching schemas."""
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

from src.app.models.matching import MatchingStatus, MatchingType


# Base schemas
class MatchingBase(BaseModel):
    """Base matching schema."""

    matching_type: MatchingType = MatchingType.AUTO
    matching_reason: str | None = None


class MatchingCreate(MatchingBase):
    """Matching creation schema (for manual matching)."""

    expert_id: UUID
    demand_id: UUID
    match_score: float | None = Field(None, ge=0, le=100)


class MatchingUpdate(BaseModel):
    """Matching update schema."""

    status: MatchingStatus | None = None
    matching_reason: str | None = None
    expert_response: str | None = None
    company_feedback: str | None = None
    company_rating: int | None = Field(None, ge=1, le=5)
    project_start_date: str | None = None
    project_end_date: str | None = None


class MatchingExpertResponse(BaseModel):
    """Expert response to matching proposal."""

    accept: bool
    response_message: str | None = None


class MatchingCompanyFeedback(BaseModel):
    """Company feedback for completed matching."""

    rating: int = Field(..., ge=1, le=5)
    feedback: str | None = None


class MatchingInDB(MatchingBase):
    """Matching schema with database fields."""

    id: UUID
    expert_id: UUID
    demand_id: UUID
    status: MatchingStatus
    match_score: float | None = None
    score_breakdown: dict | None = None
    expert_response: str | None = None
    expert_responded_at: str | None = None
    company_feedback: str | None = None
    company_rating: int | None = None
    matched_by: UUID | None = None
    project_start_date: str | None = None
    project_end_date: str | None = None
    is_active: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        """Pydantic config."""

        from_attributes = True


class Matching(MatchingInDB):
    """Matching response schema."""

    pass


class MatchingWithDetails(Matching):
    """Matching with expert and demand details."""

    expert_name: str | None = None
    expert_specialties: list[str] | None = None
    demand_title: str | None = None
    company_name: str | None = None


class MatchingList(BaseModel):
    """Matching list response schema."""

    items: list[Matching]
    total: int
    page: int
    page_size: int


class MatchingSummary(BaseModel):
    """Matching summary statistics schema."""

    total: int = 0
    proposed: int = 0
    accepted: int = 0
    rejected: int = 0
    in_progress: int = 0
    completed: int = 0


# Auto-matching schemas
class AutoMatchRequest(BaseModel):
    """Auto-matching request schema."""

    demand_id: UUID
    max_candidates: int = Field(5, ge=1, le=20)
    min_score: float = Field(60.0, ge=0, le=100)


class MatchCandidate(BaseModel):
    """Match candidate schema."""

    expert_id: UUID
    expert_name: str
    match_score: float
    score_breakdown: dict
    specialties: list[str] | None = None
    qualification_status: str


class AutoMatchResponse(BaseModel):
    """Auto-matching response schema."""

    demand_id: UUID
    candidates: list[MatchCandidate]
    total_candidates: int


# Enhanced matching schemas (Sprint 6)
class MatchScoreBreakdown(BaseModel):
    """Detailed score breakdown for intelligent matching."""

    specialty: float = Field(..., description="전문분야 일치 점수 (40%)")
    qualification: float = Field(..., description="자격 검증 점수 (15%)")
    career: float = Field(..., description="경력 점수 (15%)")
    evaluation: float = Field(..., description="평가 점수 (20%)")
    availability: float = Field(..., description="가용성 점수 (10%)")


class RecommendedCandidate(BaseModel):
    """Recommended candidate with intelligent scoring."""

    expert_id: UUID
    expert_name: str
    total_score: float = Field(..., ge=0, le=100)
    score_breakdown: MatchScoreBreakdown
    recommendation_reasons: list[str]
    specialties: list[str] | None = None
    qualification_status: str


class RecommendRequest(BaseModel):
    """Request for intelligent recommendations."""

    demand_id: UUID
    top_n: int = Field(10, ge=1, le=50, description="최대 추천 수")
    min_score: float = Field(50.0, ge=0, le=100, description="최소 점수 임계값")


class RecommendResponse(BaseModel):
    """Response with intelligent recommendations."""

    demand_id: UUID
    demand_title: str | None = None
    candidates: list[RecommendedCandidate]
    total_candidates: int
    algorithm_version: str = "v1.0"


class CompatibilityCheckResponse(BaseModel):
    """Compatibility check between expert and demand."""

    expert_id: str
    demand_id: str
    total_score: float
    score_breakdown: MatchScoreBreakdown
    recommendation: str  # HIGHLY_RECOMMENDED, RECOMMENDED, POSSIBLE, NOT_RECOMMENDED
    recommendation_text: str  # Korean text
    reasons: list[str]
    details: dict


class MatchingAnalytics(BaseModel):
    """Matching analytics and statistics."""

    status_distribution: dict[str, int]
    success_rate: float
    average_match_score: float
    total_active_matchings: int
    total_completed: int
    top_matched_experts: list[dict]

File: frontend/src/types/matching.ts
/**
 * Matching types for expert-demand matching system.
 *
 * Aligned with backend schemas: src/app/schemas/matching.py
 */

// Matching status enum (aligned with backend MatchingStatus)
export type MatchingStatus =
  | 'PROPOSED'
  | 'ACCEPTED'
  | 'REJECTED'
  | 'IN_PROGRESS'
  | 'COMPLETED'
  | 'CANCELLED';

// Matching type enum
export type MatchingType = 'AUTO' | 'MANUAL';

// Base matching interface
export interface Matching {
  id: string;
  expert_id: string;
  demand_id: string;
  matching_type: MatchingType;
  status: MatchingStatus;
  match_score: number | null;
  score_breakdown: Record<string, unknown> | null;
  matching_reason: string | null;
  expert_response: string | null;
  expert_responded_at: string | null;
  company_feedback: string | null;
  company_rating: number | null;
  matched_by: string | null;
  project_start_date: string | null;
  project_end_date: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

// Matching with additional details
export interface MatchingWithDetails extends Matching {
  expert_name?: string;
  expert_specialties?: string[];
  demand_title?: string;
  company_name?: string;
}

// Matching list response
export interface MatchingList {
  items: Matching[];
  total: number;
  page: number;
  page_size: number;
}

// Matching summary statistics
export interface MatchingSummary {
  total: number;
  proposed: number;
  accepted: number;
  rejected: number;
  in_progress: number;
  completed: number;
}

// Create matching request
export interface MatchingCreate {
  expert_id: string;
  demand_id: string;
  matching_type?: MatchingType;
  matching_reason?: string;
  match_score?: number;
}

// Update matching request
export interface MatchingUpdate {
  status?: MatchingStatus;
  matching_reason?: string;
  expert_response?: string;
  company_feedback?: string;
  company_rating?: number;
  project_start_date?: string;
  project_end_date?: string;
}

// Expert response to matching proposal
export interface MatchingExpertResponse {
  accept: boolean;
  response_message?: string;
}

// Company feedback for completed matching
export interface MatchingCompanyFeedback {
  rating: number; // 1-5
  feedback?: string;
}

// Auto-match request
export interface AutoMatchRequest {
  demand_id: string;
  max_candidates?: number;
  min_score?: number;
}

// Basic match candidate (for auto-match)
export interface MatchCandidate {
  expert_id: string;
  expert_name: string;
  match_score: number;
  score_breakdown: Record<string, {
    score: number;
    max: number;
    description: string;
  }>;
  specialties: string[] | null;
  qualification_status: string;
}

// Auto-match response
export interface AutoMatchResponse {
  demand_id: string;
  candidates: MatchCandidate[];
  total_candidates: number;
}

// ==============================================================================
// Sprint 6: Enhanced Matching Types (Intelligent Recommendations)
// ==============================================================================

// Detailed score breakdown for intelligent matching
export interface MatchScoreBreakdown {
  specialty: number;      // 40% weight
  qualification: number;  // 15% weight
  career: number;         // 15% weight
  evaluation: number;     // 20% weight
  availability: number;   // 10% weight
}

// Recommended candidate with intelligent scoring
export interface RecommendedCandidate {
  expert_id: string;
  expert_name: string;
  total_score: number;
  score_breakdown: MatchScoreBreakdown;
  recommendation_reasons: string[];
  specialties: string[] | null;
  qualification_status: string;
}

// Recommend request
export interface RecommendRequest {
  demand_id: string;
  top_n?: number;     // default: 10, max: 50
  min_score?: number; // default: 50.0
}

// Recommend response
export interface RecommendResponse {
  demand_id: string;
  demand_title: string | null;
  candidates: RecommendedCandidate[];
  total_candidates: number;
  algorithm_version: string;
}

// Recommendation level type
export type RecommendationLevel =
  | 'HIGHLY_RECOMMENDED'
  | 'RECOMMENDED'
  | 'POSSIBLE'
  | 'NOT_RECOMMENDED';

// Compatibility check response
export interface CompatibilityCheckResponse {
  expert_id: string;
  demand_id: string;
  total_score: number;
  score_breakdown: MatchScoreBreakdown;
  recommendation: RecommendationLevel;
  recommendation_text: string;
  reasons: string[];
  details: Record<string, unknown>;
}

// Matching analytics
export interface MatchingAnalytics {
  status_distribution: Record<MatchingStatus, number>;
  success_rate: number;
  average_match_score: number;
  total_active_matchings: number;
  total_completed: number;
  top_matched_experts: Array<{
    expert_id: string;
    match_count: number;
  }>;
}

// Helper: Get recommendation color based on level
export function getRecommendationColor(level: RecommendationLevel): string {
  switch (level) {
    case 'HIGHLY_RECOMMENDED':
      return 'green';
    case 'RECOMMENDED':
      return 'blue';
    case 'POSSIBLE':
      return 'yellow';
    case 'NOT_RECOMMENDED':
      return 'red';
    default:
      return 'gray';
  }
}

// Helper: Get matching status label in Korean
export function getMatchingStatusLabel(status: MatchingStatus): string {
  const labels: Record<MatchingStatus, string> = {
    PROPOSED: '제안됨',
    ACCEPTED: '수락됨',
    REJECTED: '거절됨',
    IN_PROGRESS: '진행중',
    COMPLETED: '완료',
    CANCELLED: '취소됨',
  };
  return labels[status] || status;
}

// Helper: Get matching status color
export function getMatchingStatusColor(status: MatchingStatus): string {
  const colors: Record<MatchingStatus, string> = {
    PROPOSED: 'blue',
    ACCEPTED: 'green',
    REJECTED: 'red',
    IN_PROGRESS: 'yellow',
    COMPLETED: 'green',
    CANCELLED: 'gray',
  };
  return colors[status] || 'gray';
}
